lib/nuget-parser/nuspec-parser.ts

```diff
@@ -8,80 +8,126 @@ const debug = debugModule('snyk');
 
 const targetFrameworkRegex = /([.a-zA-Z]+)([.0-9]+)/;
 
-export async function parseNuspec(dep, targetFramework) {
-  return Promise.resolve()
-    .then(() => {
-      const nupkgPath = path.resolve(
-        dep.path,
-        dep.name + '.' + dep.version + '.nupkg',
+async function loadNuspecFromAsync(dep) {
+  const nupkgPath = path.resolve(
+    dep.path,
+    dep.name + '.' + dep.version + '.nupkg',
+  );
+
+  //just in case, does the code expect null to be returned on any error here?
+  if (!fs.existsSync(nupkgPath)) return null;
+
+  const nupkgData = fs.readFileSync(nupkgPath);
+  const nuspecZipData: any = await JSZip.loadAsync(nupkgData);
+
+  const nuspecFiles = Object.keys(nuspecZipData.files).filter(file => {
+    return path.extname(file) === '.nuspec';
+  });
+
+  if (!nuspecFiles) {
+    debug(`failed to read nupkg file from: ${nupkgPath}`);
+    return null;
+  }
+
+  if (!nuspecZipData) {
+    debug(`failed to open nupkg file as an archive from: ${nupkgPath}`);
+    return null;
+  }
+
+  return await nuspecZipData.files[nuspecFiles[0]].async('text');
+}
+
+export async function actuallyParsedNuspec(
+  nuspecContent: any,
+  targetFramework,
+  depname,
+) {
+  const result = await parseXML.parseStringPromise(nuspecContent);
+  let ownDeps: any = [];
+  // We are only going to check the first targetFramework we encounter
+  // in the future we may want to support multiple, but only once
+  // we have dependency version conflict resolution implemented
+  if (!result.package?.metadata) {
+    throw new Error(
+      'This is an invalid nuspec file. Package or Metadata xml section is missing. This is a required element. See https://docs.microsoft.com/en-us/nuget/reference/nuspec',
+    );
+  }
+
+  //just in case, this shoud not happen
+  if (!Array.isArray(result.package.metadata)) {
+    throw new Error(
+      'This is an invalid nuspec file; the metadata tag is supposed to be a collection of objects but it is not!',
+    );
+  }
+
+  result.package.metadata.forEach(metadata => {
```

Comment:
It feels weird to throw an error inside of a callback, wdyt about changing it to:
```
for (let metadata of result.package.metadata) {
  ...
}
```
?